src/action.py

```diff
@@ -3,30 +3,69 @@
 import click
 import os
 from models.config import Config
+from collections import defaultdict
+from itertools import chain
 
 class ClickGlob:
-    def command(click_group, namespace):
-        def __glob_actions(namespace):
-            parts = ['/var/lib/adminware/tools',
-                     namespace, '*/config.yaml']
-            paths = glob.glob(os.path.join(*parts))
-            return list(map(lambda x: Action(x), paths))
+    def __glob_actions(namespace):
+        parts = ['/var/lib/adminware/tools',
+                 namespace, '*/config.yaml']
+        paths = glob.glob(os.path.join(*parts))
+        return list(map(lambda x: Action(x), paths))
 
+    def command(click_group, namespace):
         # command_func is either run_open or run_batch, what this is decorating
         def __command(command_func):
-            actions = __glob_actions(namespace)
+            actions = ClickGlob.__glob_actions(namespace)
             for action in actions:
                 action.create(click_group, command_func)
 
         return __command
 
+    def command_family(click_group, namespace):
+        def __command_family(command_func):
+            actions = ClickGlob.__glob_actions(namespace)
+            families = []
+            for action in actions:
+                if action.get_families(): families += [action.get_families()]
+            families_dict = __combine_dicts(families)
+            for key in families_dict.keys():
+                __create_option(click_group, command_func, key, families_dict[key])
+
+        # when want to combine the family dictionaries - while concatenating duplicate values
+        def __combine_dicts(dicts):
+            families_dict = defaultdict(list)
+            for single_dict in dicts:
+                for k, v in single_dict.items():
+                    families_dict[k].append(v)
+            # sort each command family
+            for key in families_dict:
+                families_dict[key].sort()
+            return families_dict
+
+        def __create_option(click_group, command_func, family, commands):
+            def action_family_func():
+                return command_func(family, commands)
+            action_family_func.__name__ = (family)
+            action_family_func = click_group.command(
+                    help='Run the command family \'{}\''.format(family)
+                    )(action_family_func)
 
+        return __command_family
 
 class Action:
     def __init__(self, path):
         self.path = path
         self.config = Config(self.path)
 
+    def get_families(self):
+        # return a dict of each family this action is in, pointing to this action's name
+        families = self.config.families()
+        family_dict = {}
```

Comment:
Another `family_dict`? Can all the dictionary creation be made within the `Action` object? 

From my understanding, would it be possible to have a `action.is_in_family?(name)` method? Then the runner can then just loop through all the actions and check if it belongs to the family.

This might remove some of the dictionary combining? The `actions` will still need to be looped over to create the commands, but this should only require an array.

Reply:
As per above, doesn't Action refer to a single tool/cmd? Please correct me if I'm wrong. Also I understand that the amount of dicts is overboard - it was with the intention of making the final result (a dict of each command family with values referring to the commands in that family) as clear as possible
As for the action.is_in_family?(name) function - Yeah that's a good shout actually but it'd have to be called externally (probably from the ActionFamily class) somewhere that a list of all command families has already been generated. I'll see how I get on with refactoring this into two classes (as mentioned above) and report back